metadata/private/rules/package_metadata.bzl

```diff
@@ -0,0 +1,85 @@
+"""Declares rule `package_metadata`."""
+
+load("//private/providers:package_attribute_info.bzl", "PackageAttributeInfo")
+load("//private/providers:package_metadata_info.bzl", "PackageMetadataInfo")
+
+def _package_metadata_impl(ctx):
+    attributes = [a[PackageAttributeInfo] for a in ctx.attr.attributes]
+
+    metadata = ctx.actions.declare_file("{}.package-metadata.json".format(ctx.attr.name))
+
+    ctx.actions.write(
+        output = metadata,
+        content = json.encode({
+            "attributes": {a.kind: a.metadata.path for a in attributes},
+            "label": str(ctx.label),
+            "purl": ctx.attr.purl,
+        }),
+    )
+
+    return [
+        DefaultInfo(
+            files = depset(
+                direct = [
+                    metadata,
+                ],
+            ),
+        ),
+        PackageMetadataInfo(
+            metadata = metadata,
+            files = depset(
+                direct = [
+                    metadata,
+                ],
+                transitive = [a.files for a in attributes],
+            ),
+        ),
+    ]
+
+_package_metadata = rule(
+    implementation = _package_metadata_impl,
+    attrs = {
+        "attributes": attr.label_list(
+            mandatory = False,
+            doc = """
+""".strip(),
+            providers = [
+                PackageAttributeInfo,
+            ],
+        ),
+        "purl": attr.string(
+            mandatory = True,
+            doc = """
+The [PURL](https://github.com/package-url/purl-spec) uniquely identifying this
+package.
+""".strip(),
+        ),
+    },
+    provides = [
+        PackageMetadataInfo,
+    ],
+    doc = """
+Rule for declaring `PackageMetadataInfo`, typically of a `bzlmod` module.
+""".strip(),
+)
+
+# TODO(yannic): Remove wrapper when Starlark rules support `deleted_attributes`.
```

Comment:
is there a ticket/issue tracking this work? Would be nice to link it and above all else I am interested myself